Add SSL Certificate Management page to Enterprise documentation (#2280)
travis-ci/docs-travis-ci-com: user/enterprise/ssl-certificate-management.md

---
title: Travis CI Enterprise SSL Certificate Management
layout: en_enterprise
---

This page contains information relating to SSL certificates and their use in Travis CI Enterprise.

> By default, Travis CI Enterprise verifies SSL for all traffic between its various components. SSL verification helps ensure that traffic is delivered between components securely. It is **strongly recommended** that you configure your installation with a valid certificate from a trusted authority.

Reviewer: 'various component' -> we could be specific here.

Author: Addressed in 51e00208.

If you have questions about a specific section or have an issue that is not covered then please contact us at [enterprise@travis-ci.com](mailto:enterprise@travis-ci.com) for assistance.

## Using Self-signed Certificates

We strongly recommend that you use a valid certificate from a trusted certificate authority when configuring your installation. That said, we recognize that there are various scenarios in which using a self-signed certificate might make more sense for your organization.

> In all cases of using self-signed certificates you must [disable SSL certificate validation](#disabling-ssl-verification) in your settings. This can be done after installation is complete.

> > If you decide at a later time to upload your own custom certificate then this can be done at any time via the [admin dashboard](#how-to-add-or-update-certificates-after-installation).

Author: I believe that this could be open to interpretation and that this very point results in a lot of support questions for us. 🤔 

I have updated this line to remove the 'your own custom' language to make it more generic and I'll DM you to discuss the rest.

> > If you decide at a later time to upload your own custom certificate then this can be done at any time via the [admin dashboard](#how-to-add-or-update-certificates-after-installation).

Author: Addressed in 51e00208.

> #### Trial Users

Author: Addressed in 51e00208.

> If you are using a trial license to evaluate Travis CI Enterprise then going through the effort of obtaining a trusted certificate might not make sense at this time. For this scenario we offer the ability to use self-signed certificates that are generated by Travis CI Enterprise itself.

Author: Addressed in 51e00208.

To do this you can select the "Use Self-Signed Cert" option when prompted to provide a custom SSL certificate during installation.

#### Organizations with an internally managed certificate authority

Some organizations manage their own internal certificate authority. An internal authority may or may use a root certificate provided by a [trusted root authority](https://en.wikipedia.org/wiki/Certificate_authority).

> #### Without a root certificate from a trusted CA

Author: Addressed in 51e00208.

Certificates created by an internal CA can be uploaded during installation in the same way as any other certificate.

Please be aware that you will also need to be sure to follow the [instructions on disabling SSL verification](#disabling-ssl-verification) below in order to use Travis CI Enterprise.

#### With a root certificate from a trusted CA

If your organization's internal CA does use a root certificate purchased from a trusted CA then the above will not apply. You will not need to disable SSL verification and can proceed as if you are using a certificate directly from a trusted CA.

> In some cases it may be necessary to include the intermediate certificate chain in the settings in order for SSL verification to work correctly. The reasons for this can vary depending on your internal setup and your root certificate provider. Please see the section below on [how to set intermediate certificate chains](#using-intermediate-certificate-chains) for instructions.

### Disabling SSL Verification

> SSL verification helps ensure that all traffic between Travis CI Enterprise components is protected. Disabling SSL verification has serious implications!

> You can disable SSL verification across Travis CI Enterprise via the following steps:

Reviewer: ```suggestion
To disable SSL verification in Travis CI Enterprise:
```

Author: Addressed in 51e00208.

> - Log into your dashboard at `https://<your-travis-ci-enterprise-domain>:8800`.

Reviewer: Use a numbered list here, for steps in order.

Reviewer: applies throughout

Author: Addressed in 51e00208.

- Click on 'Settings' in the topmost menu.
- Click 'Advanced Settings' in the left menu (or scroll down).
- Check the box for 'Disable SSL cert verification'.
- Click 'Save' at the bottom of the page.

> **Please note that this will require a restart of Travis CI Enterprise in order to take affect and may cause brief downtime.**

Reviewer: Fold this into the last step?

Reviewer: applies throughout

Author: Addressed in 51e00208.

- Click 'Save' at the bottom of the page.

**Please note that this will require a restart of Travis CI Enterprise in order to take affect and may cause brief downtime.**

Upon restart you can then verify whether SSL verification is now working. If you are still experiencing trouble even after setting the chain then please contact us at [enterprise@travis-ci.com](mailto:enterprise@travis-ci.com) for assistance.

## Using a Let's Encrypt SSL Certificate

Travis CI Enterprise works well together with a Let's Encrypt SSL Certificate. To obtain one for your domain, please follow the steps below.

What you need:

1. An email address Let's Encrypt can send emails to. This will be used to notify about urgent renewal and security notices.
2. A domain name under which your installation is available (we're using `travis.example.com` in this guide).

**Please note: This change will cause downtime for your system,for this reason, we recommend to perform this process in a maintenance window.**

To obtain an SSL certificate we'll be using [certbot](https://certbot.eff.org/#ubuntutrusty-other). Certbot is available as an Ubuntu package.

> Please login to your platform machine via SSH and run the following steps to install certbot:

Reviewer: log in for a verb I think

Author: Addressed in 51e00208.

```
$ sudo apt-get update
$ sudo apt-get install software-properties-common
$ sudo add-apt-repository ppa:certbot/certbot
$ sudo apt-get update
$ sudo apt-get install certbot
```

`certbot` offers multiple ways to obtain a certificate. We'll pick the `temporary webserver` option since it doesn't require any additional configuration. The only prerequisite though is that the Travis CI Enterprise container has to be stopped so that webserver can bind to port 443 properly.

> ***To start***, please stop Travis CI Enterprise:

Reviewer: :-D

I'm kinda confused as to which numbered steps are which here. If you indent paragraphs to further then the base text, you can include them as part of the numbered list (also code blocks) 

1.  Like so
       
       ```
       Test
       ```
2. Test

Reviewer: How about something like

1. Stop travis CI enterprise by running ...

2. Run the interactive certificate process by.... , selecting `X and then Y`

Author: I agree with this these suggestions but I would like to leave this as a followup. This is already present today, I simply copy/pasted it from the other page since we now have a central page for SSL-related material. I will add a TODO to my list to revisit this immediately following this PR.
